### PR TITLE
Selectable library loaders, including link-time

### LINKED ISSUE #16: [Request] cURL compile-time option
Heya,

I spoke with @MikuAuahDark briefly about this on Discord. It would be nice if lua-https could be statically compiled instead of dynamically linked. It makes it a hassle to update my fork for homebrew usage. I'm hoping I can make a pull request at some point, but I'm not great at dealing with fixing something like this, unless I use a ton of `#ifdef` markings which can look ugly.

Statically linking would let me at least only need to update the config.h file (and rename main.cpp to https.cpp to not conflict names just as a safety measure).

### PR DESCRIPTION
**Pull request by bartbes**
Resolves #16, as per my proposal there. Alternative to #22.

I wanted to do the first commit for OpenSSL 3 support anyway, and it provided an alternative solution for #16. It does feel slightly more invasive than I'd like.

@TurtleP can you check whether this works for you?

### CODE REVIEW

**Comment by TurtleP** [comment]
Alright, so I have set up my CMakeLists.txt to pull in your branch via FetchContent @bartbes (and set up LÖVE Potion as needed), but I am running into the following error (which might be where I need to patch things on my end):

```
G:/GitHub/C++/lovepotion/build/_deps/lua-https-src/src/common/HTTPS.cpp:70:
(.text._Z41__static_initialization_and_destruction_0v+0x4): 
undefined reference to `LibraryLoader::GetCurrentProcessHandle()'
```

I've set `LIBRARY_LOADER` to `linktime` and `USE_CURL_BACKEND` to `ON`. I don't think I missed any files I need, either:
```cmake
add_library(lua-https STATIC
    ${lua-https_SOURCE_DIR}/src/common/config.h
    ${lua-https_SOURCE_DIR}/src/common/LibraryLoader.h
    ${lua-https_SOURCE_DIR}/src/common/Connection.h
    ${lua-https_SOURCE_DIR}/src/common/ConnectionClient.h
    ${lua-https_SOURCE_DIR}/src/common/HTTPRequest.cpp
    ${lua-https_SOURCE_DIR}/src/common/HTTPRequest.h
    ${lua-https_SOURCE_DIR}/src/common/HTTPS.cpp
    ${lua-https_SOURCE_DIR}/src/common/HTTPS.h
    ${lua-https_SOURCE_DIR}/src/common/HTTPSClient.cpp
    ${lua-https_SOURCE_DIR}/src/common/HTTPSClient.h
    ${lua-https_SOURCE_DIR}/src/common/PlaintextConnection.cpp
    ${lua-https_SOURCE_DIR}/src/common/PlaintextConnection.h
    ${lua-https_SOURCE_DIR}/src/generic/LinktimeLibraryLoader.cpp
    ${lua-https_SOURCE_DIR}/src/generic/CurlClient.cpp
    ${lua-https_SOURCE_DIR}/src/generic/CurlClient.h
    ${lua-https_SOURCE_DIR}/src/generic/OpenSSLConnection.cpp
    ${lua-https_SOURCE_DIR}/src/generic/OpenSSLConnection.h
    ${lua-https_SOURCE_DIR}/src/lua/main.cpp
)

set(LIBRARY_LOADER "linktime")
set(USE_CURL_BACKEND ON)
```

**Comment by bartbes** [comment]
I see you’ve set LIBRARY_LOADER, but that is a cmake-only variable. Is the define HTTPS_LIBRARY_LOADER_LINKTIME set? Otherwise LinktimeLibraryLoader.cpp will be ifdeffed out.

**Comment by TurtleP** [comment]
Ah, gotcha.. didn't see that part of it. It built the 3DS binary, will check if this works shortly.

**Comment by TurtleP** [comment]
Alright, it *kind of* works. I got the following error, which I vaguely remember how I solved it last time, but I'm not 100% sure. If memory serves it failed the connection test. I might be misremembering, though?

```
[LOVE] Error: [string "main.lua"]:5: No applicable HTTPS implementation found
stack traceback:
        [string "boot.lua"]:423: in function <[string "boot.lua"]:419>
        [C]: in function 'assert'
        [string "main.lua"]:5: in main chunk
        [C]: in function 'require'
        [string "boot.lua"]:391: in function <[string "boot.lua"]:143>
        [C]: in function 'xpcall'
        [string "boot.lua"]:433: in function <[string "boot.lua"]:426>
        [C]: in function 'xpcall'       1`
```

The version of cURL we have uses certificates that are placed on the SD Card, but I doubt it's expired by now, but I can also check that.

**Comment by TurtleP** [comment]
Digging further, it's not even enabling cURL as a backend:

```cpp
HTTPSClient::Reply request(const HTTPSClient::Request& req)
{
    std::printf("request\n");
    for (size_t i = 0; clients[i]; ++i)
    {
        HTTPSClient& client = *clients[i];
        std::printf("client.valid()? %d\n", client.valid());
        if (client.valid())
            return client.request(req);
    }
    throw std::runtime_error("No applicable HTTPS implementation found");
}
```

This doesn't even print inside the for loop. I'm sure I missed something stupid, going to check the source a bit closer to see what it might be.

**Comment by TurtleP** [comment]
Okay, got it to work with the following:

```cmake
set(LIBRARY_LOADER "linktime")

add_library(lua-https STATIC
    ${lua-https_SOURCE_DIR}/src/common/config.h
    ${lua-https_SOURCE_DIR}/src/common/LibraryLoader.h
    ${lua-https_SOURCE_DIR}/src/common/Connection.h
    ${lua-https_SOURCE_DIR}/src/common/ConnectionClient.h
    ${lua-https_SOURCE_DIR}/src/common/HTTPRequest.cpp
    ${lua-https_SOURCE_DIR}/src/common/HTTPRequest.h
    ${lua-https_SOURCE_DIR}/src/common/HTTPS.cpp
    ${lua-https_SOURCE_DIR}/src/common/HTTPS.h
    ${lua-https_SOURCE_DIR}/src/common/HTTPSClient.cpp
    ${lua-https_SOURCE_DIR}/src/common/HTTPSClient.h
    ${lua-https_SOURCE_DIR}/src/common/PlaintextConnection.cpp
    ${lua-https_SOURCE_DIR}/src/common/PlaintextConnection.h
    ${lua-https_SOURCE_DIR}/src/generic/LinktimeLibraryLoader.cpp
    ${lua-https_SOURCE_DIR}/src/generic/CurlClient.cpp
    ${lua-https_SOURCE_DIR}/src/generic/CurlClient.h
    ${lua-https_SOURCE_DIR}/src/generic/OpenSSLConnection.cpp
    ${lua-https_SOURCE_DIR}/src/generic/OpenSSLConnection.h
    ${lua-https_SOURCE_DIR}/src/lua/main.cpp
)

target_compile_definitions(lua-https PRIVATE
    HTTPS_LIBRARY_LOADER_LINKTIME
    HTTPS_BACKEND_CURL
)

# link curl
pkg_check_modules(libcurl REQUIRED IMPORTED_TARGET libcurl)
target_link_libraries(lua-https PRIVATE PkgConfig::lua51 PkgConfig::libcurl)
```

It's likely because of the way I'm doing this that lua-https isn't happy with certain things. If there's a better way for me to do this, let me know.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -2,6 +2,7 @@ cmake_minimum_required (VERSION 3.13)
 
 ### Basic compilation settings
 set (CMAKE_POSITION_INDEPENDENT_CODE TRUE)
+set (CMAKE_CXX_STANDARD 11)
```
Comment (MikuAuahDark) [inline src/CMakeLists.txt:5]: I think it's safe to require C++14 as LOVE 12 already did require C++17. Well, can always bump it up if needed.

**Review by MikuAuahDark — APPROVED** [review]
In overall, I think it looks good to me.

**Pull request merged**